gax/src/main/java/com/google/api/gax/batching/v2/BatchingSettings.java

```diff
@@ -0,0 +1,120 @@
+ * request being queued up with other requests. When any of the set thresholds are reached, the
+ * queued up requests are packaged together in a batch and set to the service as a single RPC. When
+ * the response comes back, it is split apart into individual responses according to the individual
+ * input requests.
+ *
+ * <p>There are several supported thresholds:
+ *
+ * <ul>
+ *   <li><b>Message Count Threshold</b>: Once this many messages are queued, send all of the
+ *       messages in a single call. The default value is 1 message.
+ *   <li><b>Request Byte Threshold</b>: Once the number of bytes in the batched request reaches this
+ *       threshold, send all of the messages in a single call, even if message count thresholds has
+ *       not been exceeded yet. The default value is 1 byte.
+ * </ul>
+ *
+ * <p>These thresholds are treated as triggers, not as limits. Thus, if a request is made with 2x
+ * the message count threshold, it will not be split apart; only one batch will be sent. Each
+ * threshold is an independent trigger and doesn't have any knowledge of the other thresholds.
+ */
+@BetaApi("The surface for batching is not stable yet and may change in the future.")
+@AutoValue
+public abstract class BatchingSettings {
+  /** Get the element count threshold to use for batching. */
+  @Nullable
+  public abstract Long getElementCountThreshold();
+
+  /** Get the request byte threshold to use for batching. */
+  @Nullable
+  public abstract Long getRequestByteThreshold();
+
+  /** Get a new builder. */
+  public static Builder newBuilder() {
+    return new AutoValue_BatchingSettings.Builder()
+        .setElementCountThreshold(1L)
+        .setRequestByteThreshold(1L);
+  }
+
+  /** Get a builder with the same values as this object. */
+  public abstract Builder toBuilder();
+
+  /**
+   * See the class documentation of {@link BatchingSettings} for a description of the different
+   * values that can be set.
+   */
+  @AutoValue.Builder
+  public abstract static class Builder {
+    /**
+     * Set the element count threshold to use for batching. After this many elements are
+     * accumulated, they will be wrapped up in a batch and sent.
+     */
+    public abstract Builder setElementCountThreshold(Long elementCountThreshold);
+
+    /**
+     * Set the request byte threshold to use for batching. After this many bytes are accumulated,
+     * the elements will be wrapped up in a batch and sent.
+     */
+    public abstract Builder setRequestByteThreshold(Long requestByteThreshold);
+
+    abstract BatchingSettings autoBuild();
+
+    /** Build the BatchingSettings object. */
+    public BatchingSettings build() {
+      BatchingSettings settings = autoBuild();
+      Preconditions.checkArgument(
+          settings.getElementCountThreshold() == null || settings.getElementCountThreshold() > 0,
+          "elementCountThreshold must be either unset or positive");
```

Comment:
I think this check will allow null values for the threshold, as we kept OR check.

If we go by `non-null` then should the user set very high value just to negate/postpone its execution?

Reply:
Oops, I misread the code. But I still think that these should be non-null. Disabling these triggers can lead to uncontrolled memory growth, so we should discourage disabling them.

Reply:
Sure, One last query: should we increase the `defaults` as when users won't provide thresholds then either these or Client specific defaults will be considered.

Reply:
I dont know what a good default would be. It's very service specific.

One half baked thought is:
We are still missing a class that can be used at the top level stub setting to bundle together the underlying callable settings, the descriptor and the batching settings. Maybe that class can look something like:

```java
class BatchingCallSettings extends UnaryCallSettings {
   class Builder {
       Builder(Descriptor) { ... }
      Builder  setElementCountThreshold(int count) { ... }
     ....
}
```

This way the user never has to create a new instance of the class, ensuring that they don't unintentionally lose the service specific defaults. The part I haven't thought though is if BatchingCallSettings should contain an instance of BatchingSettings or if they should be one and the same?
What are your thoughts?

Also, @vam-google, can you share your initial thoughts here?

Reply:
Thanks alot for the quick input.
I think `BatchingCallSettings` should contain an instance of `BatchingSettings`. As in follow up PRs, we would be introducing `FlowControllers` related configuration. Those we can add into `BatchingSettings` (same as how it is now) so that the user will have related configuration in a single Builder. 
Another way to state the same thing is `setRetryableCode`, `setRetrySettings`...`setElementCountThreshold` doesn't sound consistent configuration to me.

Reply:
@igorbernstein2, @rahulKQL I don't have enough context here, but my 2 cents are the following:

**Boxed vs primitives** 

It looks like any non-positive value is already treated same way as `null` in the implementation: 
```java
settings.getElementCountThreshold() == null || settings.getElementCountThreshold() > 0`
```
At this point I don't see any advantage of using boxed values here, so I would choose primitives (with <=0 values treated as "not set", or even something like `<0` as 'not set' and `0` as 'use default') 


**BatchingCallSettings and BatchingSettings**

I'm really not sure that I understand what the quesiton is. Is it about if `BatchingCallSetings` should aggregate `BatchingSettings` or if `BatchingSettings` and `BatchingCallSettings` be merged in one class? 

Just looking at `BatchingSettings` is seems like a basic POJO class, with no funcitonality, only a set of properties. Its only purpose is to store a bunch of properties, it may contain some validation, but apart of that it does not interpret the values (i.e. has no logic). This is very consistent with the rest of the `*Settings` classes we use in gapic clients. Those settings classes usually directly correspond to the values one can found in the configs (for example .yaml files or proto annotations). 

These classes (Settings) only contain just plain data, they usually do not encapsulate more sophisticated objects, which cannot be mapped directly to a set of properties in a "yaml" file. For example check `UnaryCallSettings` implementation. Even though it aggregates other objects, those objects are still plain set of proerties (`StatusCodes` and `RetrySettings`). 

The `Descriptor` (passed in the constructor to `BatchingCallSettings`) seems like this "more sophisticated object". Maybe it is more of `BatchingCallContext`, not just settings? If it is, then aggregating settings in contex-like object is totaly fine.

And in simpler terms, if there is a good way to separate the two classes, then I would vote for separation, espectially if one of them (BatchingSettings) can stay as simple set of properties. Or another way to look at it: for example in C++ `BatchingSettings` would be a `struct`, and `BatchingCallSettings/Context` would be a `class`)

Reply:
Ok, looks like we agreement on keeping the BatchingSettings & BatchCallSettings separate
I think the general preference is to keep the thresholds primitive. 

Only possible disagreement is disabling via negative values. I would prefer to avoid special values and to keep defaults as 1 element & 1 byte. Then strongly encourage users in docs start their customizations using toBuilder on the existing BatchingSettings

Reply:
Thank you @igorbernstein2 and @vam-google for the input. 
 - I would include BatchingCallSettings in following PRs(probably after adding flow controllers), I would make sure to aggregate it then.
 - I have updated thresholds in BatchingSettings to be of primitive type.
 - I have kept the default threshold to be 1 element & 1 byte with explanation in the Javadoc.

Please let me know if this change looks good to you.